File: Python/ByTier/F/1790_F_Timofey_and_Black_White_Tree.py
# Problem: CF 1790 F - Timofey and Black-White Tree
# https://codeforces.com/contest/1790/problem/F

"""
Algorithm: 
- Build a tree from input edges
- Perform DFS to find parent relationships and compute distances from initial black node (c0)
- For each operation:
    - Traverse up the tree from newly colored black vertex to root
    - Update minimum distances from current black vertices
    - Track the minimum distance between all pairs of black vertices

Time Complexity: O(n) for each test case, as each node is visited at most once per operation in the worst case.
Space Complexity: O(n) for storing tree structure, parent array, and distance array.

This approach efficiently maintains the minimum distance between all pairs of black vertices by propagating updates from newly colored nodes up to their ancestors.
"""

import os
import sys
from io import BytesIO, IOBase

INF = 10**9


def read_ls(func=int):
    return list(map(func, input().split(" ")))


def solve():
    n, c0 = read_ls()
    c0 -= 1 # Convert to 0-based indexing
    c = read_ls(lambda x: int(x) - 1) # Convert to 0-based indexing
    tree = [[] for _ in range(n)]
    for _ in range(n - 1):
        i, j = read_ls(lambda x: int(x) - 1)
        tree[i].append(j)
        tree[j].append(i)
    res = []
    parent = [-1] * n
    cur = INF
    dis = [INF] * n

    def get_parent(node):
        st = [node]
        while st:
            node = st.pop()
            for nb in tree[node]:
                if nb != parent[node]:
                    parent[nb] = node
                    st.append(nb)

    get_parent(c0)
    dis[c0] = 0
    for ci in c:
        v = ci
        d = 0
        # Traverse from the newly colored node up to the root
        # Update distances and check minimum distance with ancestors
        while v != -1 and d < cur:
            cur = min(cur, d + dis[v]) # Update global minimum distance
            dis[v] = min(dis[v], d)    # Update distance to current node
            v = parent[v]              # Move to parent
            d += 1                     # Increment distance
        res.append(cur)

    print(*res)


def main():
    for _ in range(int(input())):
        solve()


BUFSIZE = 8192


class FastIO(IOBase):
    newlines = 0

    def __init__(self, file):
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None

    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()

    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()

    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)


class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")


sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")
if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1790/1790_F_Timofey_and_Black_White_Tree.cpp
// Problem: CF 1790 F - Timofey and Black-White Tree
// https://codeforces.com/contest/1790/problem/F

/*
Purpose: Solve the problem of finding the minimum distance between pairs of black vertices after each operation of coloring a white vertex black in a tree.

Algorithm: 
- For each query, we maintain the minimum distances from each vertex to the set of previously colored black vertices.
- For each new black vertex, we traverse up the tree (using parent array) to update distances and find the current minimum distance between any two black vertices.
- We track the minimum distance encountered during traversal and update it accordingly.

Time Complexity: O(n) per test case, as each node is visited at most once during all operations combined.
Space Complexity: O(n), for storing the tree, parent information, and distance array.
*/

#include <stddef.h>
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;
#define endl '\n'
#define all(v) v.begin(), v.end()
#define rep(i, s, n) for (int i = s; i <= n; i++)
#define rev(i, n, s) for (int i = n; i >= s; i--)
typedef long long ll;
typedef long double ld;
typedef pair<int, int> pii;
typedef vector<vector<int>> vii;
const int inf = 1e9;
vii g;
vector<int> parent;

// DFS to compute parent of each node in the tree
void dfs(int u, int p = -1) {
  parent[u] = p;
  for (int v : g[u]) {
    if (v == p)
      continue;
    dfs(v, u);
  }
}

void solve_problem() {
  int n, c;
  cin >> n >> c;
  g.resize(n + 1);
  parent.resize(n + 1);
  for (int i = 0; i <= n; i++) {
    g[i].clear();
    parent[i] = -1;
  }
  vector<int> a(n - 1);
  for (int i = 0; i < n - 1; i++)
    cin >> a[i];
  for (int i = 0; i < n - 1; i++) {
    int u, v;
    cin >> u >> v;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  dfs(c); // Build parent array using DFS

  int ans = inf;
  vector<int> minD(n + 1, inf); // minD[i] stores the minimum distance from black vertices to vertex i
  minD[c] = 0; // distance from starting black vertex to itself is 0

  for (int i = 0; i < n - 1; i++) {
    int u = a[i]; // current white vertex to color black
    int dist = 0;
    // Traverse up the tree from u to update distances and compute current minimum distance between black vertices
    while (u != -1 && dist < ans) {
      ans = min(ans, dist + minD[u]); // update answer with distance between current vertex and closest black vertex
      minD[u] = min(minD[u], dist); // update minimum distance to this vertex
      dist += 1;
      u = parent[u]; // go to parent
    }
    cout << ans << " ";
  }
  cout << endl;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  cout << fixed << showpoint;
  cout << setprecision(15);
  int test = 1;
  cin >> test;
  while (test--)
    solve_problem();
  return 0;
}


// https://github.com/VaHiX/CodeForces/